### ISSUE TITLE
Implement Reinforcement Learning with Inhuman Feedback

### ISSUE THREAD

**Issue by abrichr** [OpenAdaptAI/OpenAdapt#393]
### Feature request

https://github.com/CarperAI/trlx

### Motivation

https://twitter.com/i/web/status/1668337702440165376

![image](https://github.com/OpenAdaptAI/OpenAdapt/assets/774615/d04a4f61-8e3b-4a95-a6be-0aa8150deb87)

https://www.youtube.com/watch?v=DxInMGIvkp0

**Comment by LaPetiteSouris** [comment]
We can do it like this

**Step 1** We can draft out a generic script like `openadapt.utils.rf_fine_tune_models.py` which has an interface like:

```python

def tune_model(model_name: str, reward_func: Callable, extra: Any)->Model:
```
This script can:
- Fetch model from remote (HuggingFace for example)
- Fine tune models using `trlx`
- Re-upload model on HuggingFace or may be just keep it locally.

With this small util script, we can test the performace on several models and records to see if the hypothesis that Reinforcement Learning with Inhuman Feedback can actual improve the performance or not.

**Step 2** With the available script and hand, try and test to figure out what works best. If step 1 uploads the fine-tuned model in HuggingFace, then step 2 can be performed by virtually anyone in the team, at any moment. We already have HugginFace strategy if I am not mistaken. If we don't have it yet, then I can tackle the implementation later.

**Comment by atomicrichard** [comment]
Thank you @LaPetiteSouris ! Sounds like we're on the right track.

> Step 1 : Build sample dataset

I think we can implement everything before worrying about building the dataset per se. What we need is code that converts one or more recordings into a trainable dataset, e.g. in `openadapt.ml.data`. Once we've built out the rest of the pipeline we can think about increasing the size of the dataset, but keeping it small and focussed can be helpful to start. You can create one yourself just by creating a recording if you doing a repetitive action.

> Step 2 Define a metric to decide if ther is actually an improvement in the model.

Using an off the shelf package sounds good. To clarify, are you referring to https://github.com/EleutherAI/lm-evaluation-harness ? My understanding is that we want to evaluate autoregressively. I believe this means that, given a recording from time t_i to t_k, we can train the model on t_i to t_j (where t_i < t_j < t_k), evaluate its ability to predict t_j + 1, and repeat for every t_j < t_k - 1. Please confirm your understanding, and let me know if anything is unclear 🙏 

Step 3: Design a reward function.

I think we don't need this if we use a prompt-completion dataset, as per https://github.com/CarperAI/trlx#using-a-prompt-completion-dataset , but I may be wrong. Can you please confirm?

Excited to see what you come up with!

**Comment by abrichr** [comment]
Thank you @LaPetiteSouris ! Sounds like we're on the right track.

> Step 1 : Build sample dataset

I think we can implement everything before worrying about building the dataset per se. What we need is code that converts one or more recordings into a trainable dataset, e.g. in `openadapt.ml.data`. Once we've built out the rest of the pipeline we can think about increasing the size of the dataset, but keeping it small and focussed can be helpful to start. You can create one yourself just by creating a recording if you doing a repetitive action.

> Step 2 Define a metric to decide if ther is actually an improvement in the model.

Using an off the shelf package sounds good. To clarify, are you referring to https://github.com/EleutherAI/lm-evaluation-harness ? My understanding is that we want to evaluate autoregressively. I believe this means that, given a recording from time t_i to t_k, we can train the model on t_i to t_j (where t_i < t_j < t_k), evaluate its ability to predict t_j + 1, and repeat for every t_j < t_k - 1. Please confirm your understanding, and let me know if anything is unclear 🙏 

Step 3: Design a reward function.

I think we don't need this if we use a prompt-completion dataset, as per https://github.com/CarperAI/trlx#using-a-prompt-completion-dataset , but I may be wrong. Can you please confirm?

Excited to see what you come up with!

**Comment by LaPetiteSouris** [comment]
**Step 1** All good. If we are OK with using localized/miniature recording as dataset, I am all fine. This limits the scope of the task, which is in general a good things to make it happen.

**Step 2**

I guess what you describe would be done directly in step 1 (or some short), to prepare a pre-training dataset by separating the records to sections of 70-15-15 (training set-validation set-test set). I don't know yet if the [library itself](https://github.com/EleutherAI/lm-evaluation-harness) already handle the split, but it's a detail. If it is not the case I will split the dataset myself, should not be a big deal.

Let's see the performance output given by `lm-evaluation-harness` and we can have materials for discussion.

**Step 3**

If we use prompt completion dataset, we don't need reward function. But if I get the idea correctly, we want kind of like "AI to train AI" , which is different from RF with human feedback. In the article by Yam Peleg, he is using a reward function in substitution to prompt-answer dataset.

If we go with prompt dataset, we can of course pass the a few records (after processing) as prompt data for training. 
Otherwise, we have to come up with a reasonable reward function. Of course this is a bit harder.

@abrichr Thoughts ?

**Comment by abrichr** [comment]
> separating the records to sections of 70-15-15 (training set-validation set-test set)

This is typical of supervised learning tasks. Since this is semi-supervised / autoregressive, I believe we will train on the full set. If you disagree can you please point me to some documentation or literature that supports type of training split when training transformers?

> In the article by Yam Peleg, he is using a reward function in substitution to prompt-answer dataset.

Can you please clarify here what exactly he says about this and where? 🙏

**Comment by abrichr** [comment]
Rlaif: https://www.anthropic.com/index/constitutional-ai-harmlessness-from-ai-feedback

**Comment by LaPetiteSouris** [comment]
@abrichr 

You are correct. I was mistaken. Auto regressive training does not require the full split as the output is calculated from forward loop, hence we can give the full recording as dataset.

For the article on RLAIF and for your question on rewarding model, I think that here is the recap
 
<img width="911" alt="Screenshot 2023-07-21 at 09 24 23" src="https://github.com/OpenAdaptAI/OpenAdapt/assets/6369285/f7047e30-e7ab-45e7-9fc2-aa26f4a789f3">

In both cases of Yam Peleg and Anthropic, they create either a reward function or in a complex case of Anthropic, a reward models based on "guideline principle". I suggest we keep this for the next topic to limit the scope. For now, let's just:

1. Create sampling dataset from repetitive recording.
2. Evaluate original scores  of the models using `lm-evaluation-harness` .
3. Perform RF learning directly on the prompt dataset from (1), measure performance gains using step (2)
4. Repeat (2) and (3) a while until improvement is observed.

To keep it simple, let's start with a single model at the moment. I suggest trying https://github.com/OpenAdaptAI/OpenAdapt/issues/391 as the model, as it may prepare the road for this ticket as well.

WDYT ?

@abrichr

**Comment by abrichr** [comment]
@LaPetiteSouris sounds great to me!

Please implement a draft and open a PR at your earliest convenience so that we can start iterating on design 🙏

**Comment by abrichr** [comment]
Closing as part of architecture transition. Legacy frozen at v0.46.0. See PR #960.
